src/compiler/bundle/app-data-plugin.ts

```diff
@@ -52,7 +54,7 @@ export const appDataPlugin = (
       return null;
     },
 
-    load(id) {
+    load(id): d.RollupLoadHook {
```

Comment:
I've been thinking a lot on this file. 

I want to propose a couple refactoring items here that are not blocking, aren't worth adding to this PR, but worth talking about and adding a ticket to Jira if agreed upon.

Reply:
First, I think we should try to set a policy around our internal rollup plugins and their `load`/`transform` to always return either null, or the [SourceDescription](https://github.com/rollup/rollup/blob/69c099c9248d150fcb99c80c97b52adfda30fabc/src/rollup/types.d.ts#L98) object. That way, extending this code to allow source maps to other new files later down the line aren't a problem, but also if we ever want to pass along an AST without a source map, we can do just that. 

I don't necessarily agree about sometimes returning a string, or sometimes returning an object, and would love to see that as a policy for these two methods. 

I rewrote this function to be:
```typescript
    load(id) {
      const s = new MagicString(``);
      let result = null;

      if (id === STENCIL_APP_GLOBALS_ID) {
        appendGlobalScripts(globalScripts, s);
        result = {
          code: s.toString(),
          map: undefined,
        };
      } else if (id === STENCIL_APP_DATA_ID) {
        appendNamespace(config, s);
        appendBuildConditionals(config, build, s);
        appendEnv(config, s);

        result = {
          code: s.toString(),
          map: undefined,
        };
      } else if (id === config.globalScript) {
        const module = compilerCtx.moduleMap.get(config.globalScript);
        if (!module.sourceMapFileText) return { code: module.staticSourceFileText, map: null };

        const sourceMap: d.SourceMap = JSON.parse(module.sourceMapFileText);
        sourceMap.sources = sourceMap.sources.map((src) => basename(src));
        result = { code: module.staticSourceFileText, map: sourceMap };
      }

      return result;
    },
 ```

Reply:
Second, and more contentious because I don't have a deeper understanding of the flow of data - I think in an effort to remove the if statements from the plugin, we should abstract this into three separate plugins. 
1. the STENCIL_APP_GLOBALS_ID plugin
2. the STENCIL_APP_DATA_ID plugin
3. the config.globalScript plugin

That way this file can be kept small and testable. 

Meaning, we'd have three plugins with these load functions: 

app-data-globals-plugin.ts
```typescript
    load(id) {
      const s = new MagicString(``);
      let result = null;

      if (id === STENCIL_APP_GLOBALS_ID) {
        appendGlobalScripts(globalScripts, s);
        result = {
          code: s.toString(),
          map: undefined,
        };
      } 
      
      return result;
    },
```

app-data-plugin.ts
```typescript
    load(id) {
      const s = new MagicString(``);
      let result = null;

      if (id === STENCIL_APP_DATA_ID) {
        appendNamespace(config, s);
        appendBuildConditionals(config, build, s);
        appendEnv(config, s);

        result = {
          code: s.toString(),
          map: undefined,
        };
      } 
      
      return result;
    },
```

app-data-globalScript-plugin.ts
```typescript
    load(id) {
      let result = null;

      if (id === config.globalScript) {
        const module = compilerCtx.moduleMap.get(config.globalScript);
        if (!module.sourceMapFileText) return { code: module.staticSourceFileText, map: null };

        const sourceMap: d.SourceMap = JSON.parse(module.sourceMapFileText);
        sourceMap.sources = sourceMap.sources.map((src) => basename(src));
        result = { 
          code: module.staticSourceFileText, 
          map: sourceMap 
        };
      } 
      
      return result;
    },
```

I recognize this is a subset of the content of the original file, so this is naive. But this plugin is overloaded without a doubt.

Reply:
For future reference - I think passing around an AST instead of a string may improve the performance of our consumers compilation. See https://github.com/rollup/rollup/issues/782